Spring Security와 OAuth2를 활용한 인증 및 권한 관리 통합

onda0-0 opened the pull request:
https://devarium.atlassian.net/browse/BE-12

onda0-0 committed:
chore: 중복 파일 제거

onda0-0 committed:
feat: Spring security의 OAuth2기술을 최대한 활용하며 필요없는 UserDetailsService구현체 삭제

onda0-0 committed:
feat: DefaultOAuth2UserService를 CustomOAuth2UserService에서 주입받아 활용할수 있도록 빈으로 등록.

onda0-0 committed:
feat: registrationId로 OAuth2Provider를 찾는 fromString메서드 작성

onda0-0 committed:
feat: 스프링시큐리티를 최대한 활용하기 위해 DefaultOAuth2UserService를 활용한 CustomOAuth2UserService작성

onda0-0 committed:
feat: JWT 자체가 인증 정보를 포함하기에 불필요하게 데이터베이스를 조회할 필요가 없으므로 UserDetails를 이용한 불필요한 db조회 삭제

onda0-0 committed:
feat: 필요없는 의존성 및 로직 삭제

- OAuth2 인증만을 사용하고 자체 로그인 기능이 없으므로 UserDetailsService와 관련된 의존성 및 로직이 필요하지 않습니다.
-인증 성공 후 작업을 위해 OAuth2AuthenticationSuccessHandler 등록
- JWT 검증 과정에서 UserDetailsService가 활용되지 않으며, JWT만으로 사용자의 인증 상태를 확인할 수 있도록 설계되었습니다.

onda0-0 committed:
feat: OAuth2AuthenticationSuccessHandler 구현

onda0-0 committed:
feat: 모듈분리 의도에 따라 Security와 관련된 CustomUserDetails를 infra에서 api모듈로 이동, OAuth2User위임, UserEntity가 아닌 User사용

sillysillyman commented on api/src/main/java/io/devarium/api/auth/CustomOAuth2UserService.java:
1. 명시적 생성자 정의 대신 `@RequiredArgsConstructor` 사용해 주세요.
2. `UserRepository`를 주입 받는 대신 `UserService`를 주입 -> `registerNewUser` 헬퍼 메서드 제거 후 `UserService`의 `createUser` 활용은 어떤가요?
3. 디버깅을 위한 출력은 `System.out.println()` 대신 `@Slf4j` 추가 후 `log.debug()`를 사용해 주세요.

onda0-0 replied:
1. `@RequiredArgsConstructor`로 수정했고, 
2. `UserService`를 주입 및 `createUser`를 활용하는 방식으로 수정 했습니다. 
3. 알겠습니다.

sillysillyman replied:
변경된 코드는 바로 push 해주세요.

sillysillyman commented on api/src/main/java/io/devarium/api/auth/CustomUserDetails.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public String getName() {
```
`getName()`의 리턴 값을 이메일로 하는 것보다 `getEmail()`을 커스텀으로 정의하는 게 좋을 것 같습니다.

onda0-0 replied:
getEmail()을 만들고  getName() 에서 getEmail()을 리턴하도록 수정했는데 그런 의미 맞으신가요?

sillysillyman replied:
그렇다기 보다는 `CustomUserDetails`에서 `email` 값을 가져올 때는 `getEmail()`을 호출하는 것으로 통일하고 사용하지 않을 `getName()`은 아래처럼 바꾸는 것을 제안한 겁니다.

```java
@Deprecated
@Override
public String getName() {
    return null;
}
```

`getName()`이 쓰이는 곳이 있을까요?

onda0-0 replied:
스프링시큐리티 내부로직에서 OAuth2User의 사용자 식별자로  자동으로 인식하기때문에 principal이 null인것으로 인식하여 예외가 발생합니다. (getUsername() 은 Userdetails의 식별자로, 동일한 문제가 발생합니다. )

getEmail을 식별자로 사용하기위해선 Spring Security는 기본적으로사용하는 DaoAuthenticationProvider외에 별도의 AuthenticationProvider를 구현한 새로운 구현체를 만들고, 해당 구현체를 등록하여 사용해야합니다. 

스프링시큐리티에서 자동으로 처리하는 로직에서만 쓰이는 메서드인데 추가적인 클래스를 만들고 이를 등록하는것보다 getEmail을 추가하여 getUsername, getName에서 getEmail을 가져와 사용하도록 하는 방법이 낫다고 판단했습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/auth/CustomUserDetails.java:
```diff
@@ -0,0 +1,65 @@
+    OAuth2User {
+
+    // 기본 생성자 (속성 없는 경우)
+    public CustomUserDetails(User user) {
```
이 생성자가 필요한 이유가 있나요? 참고로 주석의 기본 생성자는 매개변수가 없는 생성자를 뜻합니다!

onda0-0 replied:
당시에 해당 코드를 작성했던것은 혹여 attributes값이 비어있을경우 NPE를 피하기 위함이었는데, 현재는 코드를 수정한 상태로, 아예 CustomUserDetails에 User만 넘겨주기 때문에 해당 생성자는 신경쓰지 않으셔도 될것같습니다.

sillysillyman replied:
"현재는 코드를 수정한 상태로" -> 리뷰어는 push 된 코드를 기준으로 보기 때문에 push는 자주 해주시면 좋을 것 같습니다. 현재 기준으로는 변경된 코드를 볼 수 없습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/auth/OAuth2ServiceConfig.java:
OAuth2.0 설정에 추가될 것이 있는 게 아니면 빈을 `SecurityConfig`에 옮기는 것도 괜찮다고 생각합니다.

onda0-0 replied:
해당 빈을 SecurityConfig로 옮기면 순환참조가 발생하기때문에 별도의 config파일이 필요합니다. SecurityConfig로 옮겼을때 순환참조를 방지하기 위한 방법의 일환으로 DefaultOAuth2UserService를 직접 CustomOAuth2UserService내부에서 직접 생성하는 방법도 있긴하겠으나, Spring에서 관리하지 않는객체가 되므로, 별도의 config파일에서 빈을 등록하여 순환참조는 방지하고 의존성도 주입할수 있도록 하는게 좋겠다는 판단을 했습니다.

sillysillyman replied:
확인했습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/auth/handler/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,40 @@
+
+        // 응답에 추가
+        response.setHeader("Authorization", "Bearer " + accessToken);
+        response.setHeader("Refresh-Token", refreshToken);
```
정의된 `BEARER_PREFIX`를 사용해주세요. 그리고 `"Authorization"`, `"Refresh-Token"` 같은 스트링 리터럴은 사용되는 곳이 여러 곳일 경우 적절한 곳에 `public` 상수로 정의하는 게 좋을 것 같습니다.

onda0-0 replied:
JwtConstants를 활용해 정의된 상수를 사용했습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/auth/handler/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,40 @@
+
+        response.setContentType("application/json");
+        response.getWriter().write(
+            String.format("{\"accessToken\": \"%s\", \"refreshToken\": \"%s\"}", accessToken,
```
`Token` 도메인 모델을 활용해서 직렬화 해주세요.

```java
// JWT 생성
String accessToken = jwtUtil.generateAccessToken(email, authorities);
String refreshToken = jwtUtil.generateRefreshToken(email, authorities);
Token token = Token.of(accessToken, refreshToken);

...
objectMapper.writeValue(response.getWriter(), token);
```

onda0-0 replied:
수정된 부분 확인하시고 추가적인 리뷰 부탁드립니다.

sillysillyman replied:
확인했습니다!

sillysillyman commented on core/src/main/java/io/devarium/core/auth/OAuth2Provider.java:
```diff
@@ -10,4 +10,20 @@ public enum OAuth2Provider {
+            return OAuth2Provider.valueOf(registrationId.toUpperCase());
+        } catch (IllegalArgumentException ignored) {
+            // 2. provider 필드와 매핑
+            for (OAuth2Provider provider : values()) {
```
Enum `values()`의 크기가 충분히 작기 때문에 Stream을 활용하여 중첩문을 줄이고 가독성을 살리는 것은 어떨까요?

onda0-0 replied:
`return java.util.Arrays.stream(values())
                .filter(provider -> provider.getProvider().equalsIgnoreCase(registrationId))
                .findFirst()
                .orElseThrow(
                    () -> new IllegalArgumentException("Unknown provider: " + registrationId));
        }`이런식으로 변경했는데 생각하시던 방향이 맞을까요?

sillysillyman replied:
네 정확히 이런 의도였습니다.

sillysillyman committed:
chore: 불필요한 인터페이스 제거

onda0-0 committed:
chore: DefaultOAuth2UserService를 사용하게 되면서 사용하지 않는 webclient관련 파일 삭제 및 사용하지 않는 EmailPrincipal도 추가적으로 삭제

onda0-0 committed:
chore: 불필요한 설정 삭제

onda0-0 committed:
feat: OAuth2User의 구현체인 CustomUserDetails에서 메서드를 오버라이딩할때 필요한 메서드 구현

onda0-0 committed:
feat: DefaultOAuth2UserService를 사용하게 되면서 스프링에서 자동처리하는 경로로 변경

onda0-0 committed:
feat: 불필요한 메서드 제거 및 사용불가한 로직의 이동

- DefaultOAuth2UserService를 사용하게 되면서 불필요한 loginWithGoogle메서드 제거
- CustomUserDetails가 API모듈에 있어 AuthService에서 접근이 어려우므로 관련로직을 controller로 이동

onda0-0 committed:
feat: 불필요한 메서드 제거 및 사용 가능한 곳으로 로직 이동

- DefaultOAuth2UserService를 사용하게 되면서 불필요한 /google/callback 제거
- CustomUserDetails가 API모듈에 있어 AuthService에서 접근이 어려우므로 관련로직을 controller로 이동

onda0-0 committed:
feat: 소프트딜리트 상태 확인 메서드 작성

onda0-0 committed:
feat: 생성하는 객체들의 생성자 시그니처 변경반영, 스프링시큐리티 제공 경로 활용을 위해 허용경로에 "/oauth2/**" 추가

- JwtAuthenticationFilter
- OAuth2AuthenticationSuccessHandler
- 스프링시큐리티 제공 경로 활용을 위해 허용경로에 "/oauth2/**" 추가

onda0-0 committed:
feat: 필요없는 EmailPrincipal의 위임 삭제, 추가적으로 필요한 메서드 커스텀, CustomUserDetails의 생성자 시그니처 변경

- getEmail메서드를 만들고 getUsername과 getName에서 이를 호출하도록 변경
- getAttributes에서 유저의 toAttributes를 리턴하도록 변경
- getUser, getId메서드 추가.
- CustomUserDetails의 생성자 시그니처 변경

onda0-0 committed:
chore: 소프트딜리트 고려를 위한 TODO추가

onda0-0 committed:
bug: Authentication 객체에 있는 principal의 타입문제 해결 및 소프트딜리트 고려를 위한 TODO추가

- Authentication 객체에 있는 principal의 타입문제 해결을 위해 userDetails활용
- userDetails활용을 위해 customUserDetailsService구현하여 활용
- 소프트딜리트 고려를 위한 TODO추가

onda0-0 committed:
feat: UserDetailsService를 위임해 CustomUserDetailsService 구현

- JwtAuthenticationFilter에서 Authentication 객체에 있는 principal의 타입문제 해결을 위해 userDetails활용
- userDetails활용을 위해 customUserDetailsService구현하여 활용
- 소프트딜리트 고려를 위한 TODO추가

onda0-0 committed:
refactor: 인증 처리 로직 개선 및 코드 리팩토링

- 인증 객체를 CustomUserDetails로 캐스팅하도록 변경
- 이메일 추출 로직을 커스텀 메서드로 처리
- JWT 생성 방식을 TokenService 활용으로 변경
- 응답 데이터를 TokenResponse를 통해 생성
- ObjectMapper를 사용한 JSON 직렬화 적용

onda0-0 committed:
refactor: OAuth2 인증 서비스 로직 개선 및 코드 리팩토링

- UserRepository를 UserService로 대체하여 사용자 로직 분리 및 재사용성 향상
- OAuth2 사용자 정보를 OAuth2UserInfo 객체로 래핑하여 명확성 및 확장성 개선
- 이메일 유효성 검사 추가 (OAuth2 사용자 정보에 이메일이 없을 경우 예외 처리)
- 기존 사용자와 Provider 불일치 시 예외 처리 로직 유지
- 신규 사용자 등록 로직을 UserService의 `createUser` 메서드로 위임
- CustomUserDetails 생성 과정 간소화 및 객체 생성 로직 정리

onda0-0 committed:
refactor: 사용자 컨트롤러 인증 객체를 CustomUserDetails로 변경

- 인증 객체를 EmailPrincipal에서 CustomUserDetails로 교체
- getMe, updateMe, withdraw 메서드에서 CustomUserDetails를 활용하여 사용자 정보 처리
- principal 객체에서 getId() 및 getUser() 메서드 호출로 사용자 정보 조회 간소화

onda0-0 committed:
Merge remote-tracking branch 'origin/BE-12' into BE-12

sillysillyman commented on api/src/main/java/io/devarium/api/auth/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,58 @@
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
```
사소한 거긴 하지만 개발 가이드라인 상 `@Service`보다 `@RequiredArgsConstructor`를 먼저 써주시면 좋을 것 같습니다.

https://github.com/devarium/devarium-backend/wiki/%EB%B0%B1%EC%97%94%EB%93%9C-%EA%B0%9C%EB%B0%9C-%EA%B0%80%EC%9D%B4%EB%93%9C%EB%9D%BC%EC%9D%B8#33-%EC%95%A0%EB%84%88%ED%85%8C%EC%9D%B4%EC%85%98

onda0-0 committed:
feat: 응답 헤더에 토큰 추가 (리뷰반영하여 JwtConstants 활용)

- 리뷰를 반영해 JwtConstants를 활용

sillysillyman commented on api/src/main/java/io/devarium/api/auth/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,58 @@
+            throw new OAuth2AuthenticationException("Email not found in OAuth2 user info");
+        }
+
+        OAuth2UserInfo userInfo = new OAuth2UserInfo(id, email, name, picture, provider);
```
1. 정의된 정적 팩터리 메서드를 사용해 주세요.
2. `OAuth2UserInfo`에서 `id`필드가 코드에서 사용되지 않는 것 같은데 필드에 포함해야 하는지 고민해주세요.

onda0-0 replied:
1. 반영하여 수정했습니다. 
2. 현재 사용이 되지 않으나 추후 동일 이메일, 다른 provider의 조합 허용여부에 따라 필요할 가능성이 있어 현재로는 둔 상태입니다. 이후 해당 부분을 인지하고 있다가 최종적으로 동일 이메일, 다른 provider의 조합을 허용하지 않아 필요없다는 판단이 들면 지우는 방향으로 하겠습니다.

onda0-0 committed:
refactor: Stream API를 활용해 OAuth2Provider 매핑 로직 간소화

- 기존 반복문(`for`)을 Stream API로 변경하여 코드 가독성 향상

sillysillyman commented on api/src/main/java/io/devarium/api/auth/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final UserService userService;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        // UserService를 통해 사용자 정보를 조회
+        User user = userService.getUserByEmail(email);
+        //TODO: 소프트 딜리트 상태 확인 로직
+        /*if (user == null || user.isDeleted()) { // 소프트 딜리트 상태 확인
+            throw new UsernameNotFoundException("User not found or is deleted: " + email);
+        }*/
+        // User 객체를 UserDetails로 변환하여 반환
+        return new CustomUserDetails(user);
+    }
```
기존에 `loadUserByUsername` 사용하지 않고 `loadUserByEmail`을 정의해서 사용하자고 했던 것 같은데 코드가 바뀐 이유가 있을까요?

onda0-0 replied:
loadUserByEmail 를 만들어도 Spring Security가 이 메서드를 호출하지는 않습니다. Spring Security는 인증 과정에서 loadUserByUsername을 호출하기때문에 Spring Security기반의 인증과정을 따라가기 위해 해당부분을 수정했습니다. 

loadUserByEmail을 시큐리티자체에서 활용할수있게 구현하려면 이메일 기반 인증을 처리하는 커스텀 AuthenticationProvider를 구현하고 securityconfig에서 이를 등록하여 사용해야합니다. loadUserByEmail이라는 이름의 정의를 사용하기 위해 해당 과정을 거치는것이 불필요하다고 판단했습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/auth/SecurityConfig.java:
```diff
@@ -27,19 +25,10 @@ public class SecurityConfig {
+    private final CustomOAuth2UserService customOAuth2UserService;
+    private final ObjectMapper objectMapper;
+    private final TokenService tokenService;
+    private final CustomUserDetailsService customUserDetailsService;
```
`CustomOAuth2UserService`, `CustomUserDetailsService` 같은 구체적인 구현체에 의존하는 것이 아닌 인터페이스를 의존하는 게 더 좋을 것 같습니다.

onda0-0 replied:
해당부분 반영했습니다.

sillysillyman commented on api/src/main/java/io/devarium/api/controller/auth/AuthController.java:
```diff
@@ -42,8 +34,17 @@ public ResponseEntity<SingleItemResponse<TokenResponse>> refresh(
+            return ResponseEntity.noContent().build();
+        }
+
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Unauthenticated user");
```
예외 발생시켜도 될 것 같은데 어떻게 생각하시나요? 다른 엔드포인트랑 리턴 타입이 혼자 달라서 어색합니다.

onda0-0 replied:
리뷰를 반영해 예외를 발생시켜 리턴타입을 통일 시켰습니다.

sillysillyman commented on core/src/main/java/io/devarium/core/auth/service/AuthService.java:
```diff
@@ -1,61 +1,23 @@
-        } finally {
-            SecurityContextHolder.clearContext();
-        }
+    public void logout(String email) {
```
기존 코드의 시큐리티 컨텍스트 관련 로직이 사라진 이유를 설명해주시면 좋을 것 같습니다.

onda0-0 replied:
커밋에 설명해두긴했는데, CustomUserdetails의 위치가 Api에 위치하게 되면서 해당 authservice로직은 core에 있기 때문에 기존 로직에서는 CustomUserdetails에 접근해야하는데 그게 불가하여 관련로직을 controller쪽으로 옮긴 상황입니다. 현재로서 컨트롤러에서 책임이 많아지는 부분때문에 컨트롤러는 비즈니스 로직 호출 및 HTTP 응답 생성에만 집중할 수 있도록 하는 2가지 방안을 고려하고 있습니다. 

1. core에서 접근가능하도록 core에 interface를 만들어두고 해당 인터페이스를 implements하는것
2. 인증 정보 추출과 같은 로직을 별도로 분리

sillysillyman commented on core/src/main/java/io/devarium/core/domain/user/service/UserServiceImpl.java:
```diff
@@ -29,6 +29,7 @@ public User createUser(OAuth2UserInfo userInfo) {
     @Override
     public User getUser(Long userId) {
         return userRepository.findById(userId)
+            //TOdO: 소프트딜리트 고려 .filter(user -> !user.isDeleted())
```
`d`가 자기도 대문자 하고 싶대요. 🥹

onda0-0 committed:
chore: TODO에서 'd'를 'D'로 변경, 어노테이션 순서 변경

a-white-bit commented on api/src/main/java/io/devarium/api/auth/filter/JwtAuthenticationFilter.java:
```diff
@@ -37,14 +36,20 @@ protected void doFilterInternal(
                         userDetails,
                         null,
-                        userDetails.getAuthorities()
+                        jwtUtil.extractAuthorities(token)
```
`extractAuthorities(token)` 과 `getAuthorities()`를 비교해봤을 때,
`extractAuthorities()`는 JWT 파싱과 컬렉션과 스트림 연산을 포함하므로
성능적으로 `getAuthorities()`가 더 빠르다고 판단됩니다.
소프트 딜리트를 고려해서 토큰이 만들어지지 않는 경우까지 생각하면
역시 `getAuthorities()`를 사용하는 것이 좋을 것 같습니다!

onda0-0 replied:
넵 같이 고민해주셔서 감사합니다. 다시 원래 코드대로 수정해두겠습니다!

a-white-bit commented on core/src/main/java/io/devarium/core/auth/OAuth2Provider.java:
```diff
@@ -10,4 +10,19 @@ public enum OAuth2Provider {
+                .filter(provider -> provider.getProvider().equalsIgnoreCase(registrationId))
+                .findFirst()
+                .orElseThrow(
+                    () -> new IllegalArgumentException("Unknown provider: " + registrationId));
```
CustomAuthException 으로 오류 관리하면 더 좋을 것 같습니다.

a-white-bit commented on core/src/main/java/io/devarium/core/auth/OAuth2Provider.java:
```diff
@@ -10,4 +10,19 @@ public enum OAuth2Provider {
+            return OAuth2Provider.valueOf(registrationId.toUpperCase());
+        } catch (IllegalArgumentException ignored) {
+            // 2. provider 필드와 매핑
+            return java.util.Arrays.stream(values())
```
java.util.Arrays 를 import 하고 `Arrays.stream()` 사용하기

onda0-0 committed:
refactor: logout시 응답형식 통일을 위해 유효하지 않은 시도에서 예외처리

- best practice에 대한 방법을 고민했는데 일단은 응답형식을 통일시키는 부분만 반영. 추후 블랙리스트를 구현하면 로직에 어처피 변화가 생기기때문에 블랙리스트 구현시 함께 고민하여 구현.

onda0-0 committed:
refactor: OAuth2UserInfo를 정적팩토리메서드를 사용해 생성하도록 변경

onda0-0 committed:
refactor: 구현체가 아닌 인터페이스에 의존하도록 수정

- CustomOAuth2UserService -> OAuth2UserService<OAuth2UserRequest, OAuth2User>
- CustomUserDetailsService -> UserDetailsService

onda0-0 committed:
refactor: 성능 및 소프트딜리트를 고려해 권한정보를 토큰이 아닌 인증객체에서 가져오도록 수정

onda0-0 committed:
refactor: 서비스에서 제공되지않는 제공자에 대한 에러코드 커스텀 및 적용

onda0-0 committed:
refactor: java.util.Arrays 를 import 하고 Arrays.stream() 사용

onda0-0 committed:
Merge remote-tracking branch 'origin/dev' into BE-12

# Conflicts:
#	core/src/main/java/io/devarium/core/auth/exception/AuthErrorCode.java

onda0-0 committed:
refactor: dev와 병합이후 인증객체연결 수정

sillysillyman reviewed (approved):
`UserException` 핸들러만 추가해주시고 merge 해주시면 될 것 같습니다.

onda0-0 committed:
feat: Project도메인 관련 로직에 인증객체 및 유저정보 받아오도록 수정

onda0-0 committed:
feat: UserException등록

onda0-0 committed:
refactor: CustomUserDetails를 CustomUserPrincipal로 변경
